projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-checkout-flow.php

```diff
@@ -0,0 +1,213 @@
+			'woocommerceanalytics_product_view',
+			array(),
+			$product->get_id()
+		);
+	}
+
+	/**
+	 * Track the order confirmation page view
+	 */
+	public function capture_order_confirmation_view() {
+		$order_id = absint( get_query_var( 'order-received' ) );
+		if ( ! $order_id ) {
+			return;
+		}
+
+		if ( ! is_order_received_page() ) {
+			return;
+		}
+
+		$order = wc_get_order( $order_id );
+
+		$coupons     = $order->get_coupons();
+		$coupon_used = 0;
+		if ( is_countable( $coupons ) ) {
+			$coupon_used = count( $coupons ) ? 1 : 0;
+		}
+
+		if ( is_object( WC()->session ) ) {
+			$create_account = true === WC()->session->get( 'wc_checkout_createaccount_used' ) ? 'Yes' : 'No';
+		} else {
+			$create_account = 'No';
+		}
+
+		$this->record_event(
+			'woocommerceanalytics_order_confirmation_view',
+			array(
+				'coupon_used'      => $coupon_used,
+				'create_account'   => $create_account,
+				'express_checkout' => 'null', // TODO: not solved yet.
+				'guest_checkout'   => $order->get_customer_id() ? 'No' : 'Yes',
+				'oi'               => $order->get_id(),
+				'order_value'      => $order->get_total(),
+				'payment_option'   => $order->get_payment_method(),
+				'products_count'   => $order->get_item_count(),
+				'products'         => $this->format_items_to_json( $order->get_items() ),
+				'order_note'       => $order->get_customer_note(),
+				'shipping_option'  => $order->get_shipping_method(),
+			)
+		);
+	}
+
+	/**
+	 * Track the cart page view
+	 */
+	public function capture_cart_view() {
+		if ( ! is_cart() ) {
+			return;
+		}
+
+		$this->record_event(
+			'woocommerceanalytics_cart_view',
+			array_merge(
+				$this->get_cart_checkout_shared_data(),
+				array()
+			)
+		);
+	}
+
+	/**
+	 * Track the checkout page view
+	 */
+	public function capture_checkout_view() {
+		if ( ! is_checkout() ) {
+			return;
+		}
+
+		// Order received page is also a checkout page, so we need to bail out if we are on that page.
+		if ( is_order_received_page() ) {
+			return;
+		}
+
+		$this->record_event(
+			'woocommerceanalytics_checkout_view',
+			array_merge(
+				$this->get_cart_checkout_shared_data(),
+				array()
+			)
+		);
+	}
+
+	/**
+	 * Format Cart Items or Order Items to an array
+	 *
+	 * @param array|WC_Order_Item[] $items Cart Items or Order Items.
+	 */
+	protected function format_items_to_json( $items ) {
+		$products = array();
+
+		foreach ( $items as $item ) {
+			if ( $item instanceof WC_Order_Item_Product ) {
+				$product = wc_get_product( $item->get_product_id() );
+			} else {
+				$product = $item['data'];
+			}
+			$data = $this->get_product_details( $product );
+			if ( $item instanceof WC_Order_Item_Product ) {
+				$data['pq'] = $item->get_quantity();
+			} else {
+				$data['pq'] = $item['quantity'];
+			}
+			$products[] = $data;
+		}
+
+		return wp_json_encode( $products );
+	}
+
+	/**
+	 * Get Cart/Checkout page view shared data
+	 */
+	protected function get_cart_checkout_shared_data() {
```

Comment:
Why can't this be in the trait? There are a few events in `Jetpack_WooCommerce_Analytics_Universal` that would benefit from this

Reply:
Yes, I just didn't want to touch the core events data much, but I'm likely to merge everything.